### ISSUE TITLE
xcf covariance matrix of transmission field is not positive definite

### ISSUE THREAD

**Issue by londumas** [igmhub/picca#431]
Running `do_xcf.py` on  transmission files: `d_i(lambda) = T_i(lambda)/<T(lambda)>-1` for enough forests: 500,000, I get a covariance matrix that is `Warning: Matrix is not positive definite`.
This is not true for `do_cf.py` where with already 100,000 forests I get a covariance matrix that is positive definite.
This issue does not appear when the mocks are fully expanded by quickquasar.

The following plot compares the covariance matrix from transmission (Lya) and from the full expanded mocks (Lya+Continuum) for the diagonal and the off-diagonal terms, for the cross and the auto-correlation.
These plots show precisely an issue for the cross-correlation, and not for the auto-correlation:
- the diagonal terms of the "Lya" mocks are stronger than for the "Lya+Continuum",
    this is abnormal, it should be the contrary
- the correlation between pixels doesn't drop at large separation. 

Conclusion:
- The correlation in each sub-sample might not go to zero. I will have a look.

![xcf_covariance_matrix_covar](https://user-images.githubusercontent.com/3166436/40331688-653cf0c2-5d0e-11e8-88b9-3c3ffa8d7783.png)

![xcf_covariance_matrix_corr](https://user-images.githubusercontent.com/3166436/40331689-6555147c-5d0e-11e8-9b12-0261e65f935c.png)

![cf_covariance_matrix_covar](https://user-images.githubusercontent.com/3166436/40331690-65696d0a-5d0e-11e8-81fb-42368f53e5e2.png)

![cf_covariance_matrix_corr](https://user-images.githubusercontent.com/3166436/40331692-657e466c-5d0e-11e8-9b52-985f0c910e49.png)

**Comment by londumas** [comment]
If I leave the projection in the computation of the correlation function, I get a positive definite matrix:

![xcf_covariance_matrix_corr_projected](https://user-images.githubusercontent.com/3166436/40333602-8e8cd912-5d16-11e8-9b8e-2cf76c44c3d0.png)

**Comment by londumas** [comment]
Sadly, the reason is linked to something I was expecting: the mean of delta in each sub-sampling is not random. To investigate this I ran the regular xcf on the transmission files, without projection for the catalog of quasar and the random catalog.
- the first plot gives the mean all the bins of the cross-correlation in each sub-sample (1214 points): in the randoms versus in the data. We can see the correlation between the mean delta
- the second plot gives the correlation-matrix for "data", "random" and "data-random". We can see that the "data-random" is what we are used to and is positive definite
- the third plot gives a part of the correlation-matrix for "data"
- the fourth plot gives a part of the correlation-matrix for "data-random"

The reason might be that: xi = \<delta  eta\> - \<delta\>\<eta\>
Basically in `do_xcf.py`, we compute only xi = \<delta  eta\>,
the random acts as computing \<delta\>\<eta\>.
What do you think @ngbusca?

![correlation_rand_data](https://user-images.githubusercontent.com/3166436/40735210-d59734d2-63f7-11e8-9e6b-85fe1d8eb725.png)

![correlation_xcf_xcfrand_xcfminus](https://user-images.githubusercontent.com/3166436/40735208-d57c16d4-63f7-11e8-9b15-cf1bf7c7e9d7.png)



![correlationmatrix_xcf_trans](https://user-images.githubusercontent.com/3166436/40735211-d5aefc7a-63f7-11e8-8aac-dc6c24b658b4.png)

![correlationmatrix_xcfminus_trans](https://user-images.githubusercontent.com/3166436/40735213-d5caeda4-63f7-11e8-8881-c0953dca0f53.png)

**Comment by ngbusca** [comment]
@londumas what happens if you do the projection on the transmission field? Is it still not positive definite?

**Comment by londumas** [comment]
@ngbusca, With the project on the transmission filed, I get posititive definite matrices, with the correlation matrix we are used to, see comment https://github.com/igmhub/picca/issues/431#issuecomment-390805569

**Comment by londumas** [comment]
Here is the same plot as the first one but for xcf calculated on projected transmission, the correlation between random and data has totally disappeared and the values of \<delta\> are way smaller.

![correlation_rand_data_projected](https://user-images.githubusercontent.com/3166436/40741836-c937b678-6409-11e8-97df-4c4259658a19.png)

**Comment by andreufont** [comment]
This thread is really interesting. Thanks Helion!

We should think more about different estimators, and test them in mocks. Having picca is great, but it does make it harder to do code comparisons and test new estimators... In the past we had some discussion on whether we should use quasar randoms in the cross-correlation or not, and whether that made estimators more optimal. 

I still think that what we do now, i.e., stack <delta_flux> as a function of quasar position is a fine thing to do, but it is possible that it makes the covariance matrix more difficult to handle... 

By the way, what happened to the Wick covariances? I understand that you are currently computing the covarinces from sub-sampling, right? Are we still able to compute covariances based on 4-point functions? It would be interseting to see whether those are also pathological.

**Comment by ngbusca** [comment]
@andreufont the Wick covariance code is there but we haven't used it in a while. I plan to take a look at that again for our final round of eBOSS papers.

**Comment by andreufont** [comment]
Sounds great. Thanks!

**Comment by andreufont** [comment]
@londumas - The issues that @jfarr03 and I are having is that we have non-positive definite matrices when trying to use picca to measure the correlation function of the Gaussian field, the log-normal density field, the optical depth or the redshift-space optical depth. 

It would be really interesting to run these, since it would allow us to do non-trivial tests. For instance:
 - in the case of the log-normal density field we should measure (b_delta = 1, b_eta = 0)
 - in the case of the non-RSD optical depth we should measure (b_delta = 1.6, b_eta = 0)
 - in the case of the RSD optical depth we should measure (b_delta = 1.6, b_eta = 1)

**Comment by londumas** [comment]
@andreufont and @jfarr03, do you have that only for the cross or also for the auto? On the raw transmission files I don't get it for the auto, only the cross has issues.

**Comment by londumas** [comment]
The following message is an update of the previously posted ones, using the latest simulations.
The plots show the following:
 - the first plot show the covariance matrix of the cross-correlation, we can see that it is very non-diagonal for both the catalog and a random shuffle of the redshift. The covariance is not positive definite.
 - the second plot shows the correlation of \<xi\> in each healpix pixel of the correlation. Each point is one healpix and the value is the mean of the cross in this healpix.
This plots shows that the mean delta is not-zero in each healpix pixels. (This bring back bad memories from my thesis.). What is nice is that using a shuffle of the position of the quasar catches this correlation.
 - The third plot gives what we get if we remove the mean delta using the shuffled catalog from the true cross-correlation function. The covariance is diagonal and is positive definite.

![correlation_matrix_d_shuffle](https://user-images.githubusercontent.com/3166436/55519727-3a6dae00-5636-11e9-87a5-8b3c76b01c5a.png)

![correlated_delta](https://user-images.githubusercontent.com/3166436/55520147-24f98380-5638-11e9-9581-7ebccef7f6de.png)

![new_covariance_matrix](https://user-images.githubusercontent.com/3166436/55520606-55422180-563a-11e9-8df7-ff9db907c868.png)


Sending the following:
```
picca_xcf.py --drq /project/projectdirs/desi/mocks/lya_forest//london//v6.0/v6.0.0/eboss-0.0/zcat_desi_drq.fits --in-dir /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/deltas/ --z-evol-obj 1.44 --out /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/tests_no_projection/xcf_z_0_10.fits --nproc 32 --fid-Om 0.3147 --no-project --nspec 100000

picca_export.py --data /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/tests_no_projection/xcf_z_0_10.fits --out /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/tests_no_projection/xcf_z_0_10-exp.fits

picca_xcf.py --drq /project/projectdirs/desi/mocks/lya_forest//london//v6.0/v6.0.0/eboss-0.0/zcat_desi_drq.fits --in-dir /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/deltas/ --z-evol-obj 1.44 --out /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/tests_no_projection/xcf_z_0_10_shuffle_42.fits --nproc 32 --fid-Om 0.3147 --no-project --nspec 100000 --shuffle-distrib-obj-seed 42

picca_export.py --data /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/tests_no_projection/xcf_z_0_10_shuffle_42.fits --out /project/projectdirs/desi/mocks/lya_forest/picca/london//v6.0/v6.0.0/eboss-raw/tests_no_projection/xcf_z_0_10_shuffle_42-exp.fits
```

**Comment by andreufont** [comment]
Hi @londumas - I remember dicussing about randoms in cross-correlation a while ago (actually, this was first discussed by Ross O'Connell back in 2012), it might be time to properly look at this with mocks and see what is the statistical gain / loss when you take into account the randoms or not. It looks like it helps with the covariance matrix (what is great!) but I'm curious to see the end effect on BAO uncertainties. It might be worth a paper, don't you think? I could ask a student to work on this.

**Comment by londumas** [comment]
@andreufont, yes a study with a student would be good. Maybe we should change our estimator.
>statistical gain / loss when you take into account the randoms or not.

However there can't be any loss of statistical power because you can produce many realization of the random shuffles and get a near noiseless stack. I bet that 2 or 4 already are embedded into the real catalog errors.

**Comment by londumas** [comment]
This is fixed by using the random shuffle, see the following plot with 10 London mock realizations: raw and raw+continuum.
Fixed via PR https://github.com/igmhub/picca/pull/521

![image](https://user-images.githubusercontent.com/3166436/56303890-2461f200-60fa-11e9-85bb-5597692235fb.png)
